tw-tasks-core/src/main/java/com/transferwise/tasks/dao/MySqlTaskDao.java

```diff
@@ -502,20 +497,52 @@ public DeleteFinishedOldTasksResult deleteOldTasks(TaskStatus taskStatus, Durati
     int processedIdsCount = 0;
 
     for (int b = questionBuckets.length - 1; b >= 0; b--) {
-      int bucketSize = questionBuckets[b];
-      while (taskIds.size() - processedIdsCount >= bucketSize) {
+      var bucketSize = questionBuckets[b];
+
+      while (taskVersionIds.size() - processedIdsCount >= bucketSize) {
+        final int currentProcessedIdsCount = processedIdsCount;
+
+        if (tasksProperties.isParanoidTasksCleaning()) {
+          var tasksDeleteLockSql = sqlCache.computeIfAbsent(new CacheKey(LOCK_TASKS_FOR_DELETE_SQL, b),
```

Comment:
Basically lock records with SELECT FOR UPDATE.
Then take only those which versions have not changed between earlier select and this.